pkg/apis/resource/types.go

```diff
@@ -0,0 +1,466 @@
+	// freed. This ensures that resources are not leaked.
+	metav1.ObjectMeta
+
+	// Spec describes the desired attributes of a resource that then needs
+	// to be allocated. It can only be set once when creating the
+	// Claim.
+	Spec ClaimSpec
+
+	// Status describes whether the resource is available and with which
+	// attributes.
+	Status ClaimStatus
+}
+
+// ClaimSpec defines how a resource is to be allocated.
+type ClaimSpec struct {
+	// ClassName references the driver and additional parameters
+	// via the name of a Class that was created as part of the
+	// driver deployment.
+	//
+	// The apiserver does not check that the referenced class exists.
+	ClassName string
+
+	// ParametersRef references a separate object with arbitrary parameters
+	// that will be used by the driver when allocating a resource for the
+	// claim.
+	//
+	// The object must be in the same namespace as the Claim.
+	ParametersRef *ClaimParametersReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. "WaitForFirstConsumer" is the default.
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a Claim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod
+// (AllocationModeWaitForFirstConsumer). Other modes might get added in the
+// future.
+type AllocationMode string
+
+const (
+	// When a Claim has AllocationModeWaitForFirstConsumer, allocation is
+	// delayed until a Pod gets scheduled that needs the Claim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeWaitForFirstConsumer AllocationMode = "WaitForFirstConsumer"
+
+	// When a Claim has AllocationModeImmediate, allocation starts
+	// as soon as the Claim gets created. This is done without
+	// considering the needs of Pods that will use the Claim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ClaimStatus struct {
+	// DriverName is a copy of the driver name from the Class at
+	// the time when allocation started. It's necessary to enable usage of
+	// the claim by kubelet in case that the Class got removed in
+	// the meantime. It also helps the resource driver to determine whether
+	// it needs to handle a claim that got marked for deletion.
+	DriverName string
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. Nil indicates that the resource is not
+	// allocated.
+	Allocation *AllocationResult
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. Usually those are Pods, but other objects are also
+	// possible as long as they exist as resources in the apiserver. The
```

Comment:
We don't really want details of the controllers or where they live - that's all impl details

Reply:
Reduced to the first sentence.